torchgeo/datamodules/utils.py

```diff
@@ -19,6 +22,81 @@ class MisconfigurationException(Exception):
     """Exception used to inform users of misuse with Lightning."""
 
 
+class AugPipe(Module):
+    """Pipeline for applying augmentations sequentially on select data keys."""
+
+    def __init__(
+        self, augs: Callable[[dict[str, Any]], dict[str, Any]], batch_size: int
+    ) -> None:
+        """Initialize a new AugPipe instance.
+
+        Args:
+            augs: Augmentations to apply.
+            batch_size: Batch size
+        """
+        super().__init__()
+        self.augs = augs
+        self.batch_size = batch_size
+
+    def forward(self, batch: dict[str, Tensor]) -> dict[str, Tensor]:
+        """Apply the augmentation.
+
+        Args:
+            batch: Input batch.
+
+        Returns:
+            Augmented batch.
+        """
+        batch_len = len(batch["image"])
+        for bs in range(batch_len):
+            batch_dict = {
+                "image": batch["image"][bs],
+                "labels": batch["labels"][bs],
+                "boxes": batch["boxes"][bs],
+            }
+
+            if "masks" in batch:
+                batch_dict["masks"] = batch["masks"][bs]
+
+            batch_dict = self.augs(batch_dict)
+
+            batch["image"][bs] = batch_dict["image"]
+            batch["labels"][bs] = batch_dict["labels"]
+            batch["boxes"][bs] = batch_dict["boxes"]
+
+            if "masks" in batch:
+                batch["masks"][bs] = batch_dict["masks"]
+
+        # Stack images
+        batch["image"] = rearrange(batch["image"], "b () c h w -> b c h w")
+
+        return batch
+
+
+def collate_fn_detection(batch: list[dict[str, Tensor]]) -> dict[str, Any]:
```

Comment:
You definitely already explained this to me, but why can't we make these changes in the dataset instead of during collation?

Reply:
The dataset is concerned with individual samples. Our problem is not with the dataset, it's with how we represent the collection of samples. Default collate function does not work with boxes and instance masks. 

You can stack two image tensors to be of shape `(2,C,H,W)` but stacking two box tensors of shape `(N1, 4)` and `(N2, 4)` into a `(N1+N2, 4)` tensor would leave us with not knowing which boxes correspond to which image. Same for the instance masks.

So, we need to organise the samples into a list of tensors as opposed to a tensor itself.

Reply:
Gotcha. We're thinking about switching to torchvision transforms v2 someday, but not any time soon, so will keep this in mind if that makes anything simpler.